e2e/testcases/profiling_test.go

```diff
@@ -0,0 +1,608 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"fmt"
+	"path/filepath"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/gitproviders"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/taskgroup"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/e2e/nomostest/testpredicates"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core/k8sobjects"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/metadata"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TestProfilingResourcesByObjectCount loops through multiple steps. Each
+// step adds and removes incrementally more Deployments. This is useful for
+// profiling the resource requests, usage, and limits relative to the number of
+// managed objects.
+// Skipped by default, because it needs cluster autoscaling and a lot of quota.
+func TestProfilingResourcesByObjectCount(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation1, ntopts.Unstructured,
+		ntopts.ProfilingTest,
+		ntopts.WithReconcileTimeout(configsync.DefaultReconcileTimeout))
+
+	syncPath := filepath.Join(gitproviders.DefaultSyncDir, "stress-test")
+	ns := "stress-test-ns"
+
+	steps := 4
+	deploysPerStep := 500
+
+	// Make a list of RootSyncs to watch.
+	var rSyncRefs []rSyncRef
+	// All the RootSyncs use the same git repo
+	commitFn := func(nt *nomostest.NT, _ types.NamespacedName) (string, error) {
+		return nomostest.DefaultRootSha1Fn(nt, nomostest.RootSyncNN(configsync.RootSyncName))
+	}
+	// Add parent RootSync
+	rSyncRefs = append(rSyncRefs, rSyncRef{
+		GroupVersionKind: kinds.RootSyncV1Beta1(),
+		NamespacedName:   nomostest.RootSyncNN(configsync.RootSyncName),
+		SyncPath:         gitproviders.DefaultSyncDir,
+		CommitFunc:       commitFn,
+	})
+
+	for step := 1; step <= steps; step++ {
+		deployCount := deploysPerStep * step
+
+		nt.T.Logf("Adding a test namespace and %d deployments", deployCount)
+		nt.Must(nt.RootRepos[configsync.RootSyncName].Add(fmt.Sprintf("%s/ns-%s.yaml", syncPath, ns), k8sobjects.NamespaceObject(ns)))
+
+		for i := 1; i <= deployCount; i++ {
+			name := fmt.Sprintf("pause-%d", i)
+			nt.Must(nt.RootRepos[configsync.RootSyncName].AddFile(
+				fmt.Sprintf("%s/namespaces/%s/deployment-%s.yaml", syncPath, ns, name),
+				[]byte(fmt.Sprintf(`
```

Comment:
nit: pull out the definition of this pause Deployment to a function so that it's defined in one place

Reply:
done